### DISCUSSION TITLE
Select.options 依赖 ArrayTable 数据联动，数据变化 Select.options 未刷新

### DISCUSSION THREAD

**Discussion by dahui4dev** [alibaba/formily#1441] [Q&A 2.x] [answered]
## 需求

想要在页面配置下拉框的选项，所以需要在页面同时展示下拉框与表格。修改表格数据，同步修改下拉框下拉选项。

## 问题

数据变化 Select.options 未刷新，一开始以为是自己 `x-reactions` 写法有问题，查了文档时候，还是有些不太理解，试验了一下各种写法，请教了白玄，最后是 antd 的 select 的问题。

## 上代码

[https://codesandbox.io/s/young-dream-99z9n?file=/App.tsx](https://codesandbox.io/s/young-dream-99z9n?file=/App.tsx)

### 关键代码

```js
// 以下是部分代码，主要说明 “x-reactions” 的用法，完整代码看上方 codesandbox 链接。
const schema = {
  type: "object",
  properties: {
    tableItems: {
      type: "boolean",
      title: "下拉选项依赖下表",
      "x-decorator": "FormItem",
      "x-component": "Select"
      "x-reactions": {
       // 1、由于此字段依赖 `array` 字段的结果，所以这里用 `dependencies`
      //  此处 x-reactions 与 下面 2 任写一个都满足结果。
      dependencies: ["array"],
       fulfill: {
         state: {
           dataSource: "{{JSON.parse(JSON.stringify($deps[0] || []))}}"
         }
       }
      }
    },

    array: {
      type: "array",
      "x-decorator": "FormItem",
      "x-component": "ArrayTable",
      // "x-reactions": {
      //   // 2、此处如果不写条件，"array" 任何变化都会执行下方语句。 
      //   // $form.setFieldState 属于主动联动模式的一种
      //   fulfill: {
      //     // 这里纯是为了打印看 dataSource 的值
      //     run:
      //       "console.log($form.getFieldState('tableItems').dataSource);" +
      //       "$form.setFieldState('tableItems',(state)=>state.dataSource = JSON.parse(JSON.stringify($self.value || [])) )"
      //   }
      }
   }
}
```

### 结果

看打印数据，其实 select.dataSource 已经变化了，但是页面没有刷新。开始以为是 `x-reactions` 写错了

![image](https://user-images.githubusercontent.com/6160798/118621121-a997c380-b7f8-11eb-8aeb-221e6551302c.png)

### 问题原因

据作者解答

> `x-reactions` 写法是这么写的（以上关键代码，第一个 `x-reactions` 属于主动模式，第二个 `x-reactions` 属于被动模式），不过 antd select 有问题，options 更新了，没刷数据.
 

## 结论

@janryWang  像这种情况，我是只能去改 antd 的代码了吗？或者有什么其他的巧妙的方法可以实现我的需求吗？

**Comment by janryWang** [comment]
可以给antd提个bug

**Reply by dahui4dev** [reply]
我先用 antd 官网的例子试一下。然后看看提个 issues

**Reply by janryWang** [reply]
你就试试受控更新options会不会生效就知道了

**Reply by dahui4dev** [reply]
我去试一把

**Reply by dahui4dev** [reply]
确实受控组件，动态增加 options， select 不刷新。

**Comment by janryWang** [comment, marked as answer]
该问题是formily的问题，目前最新版已经修复了
